brayns/engine/model/Components.h

```diff
@@ -0,0 +1,158 @@
+/* Copyright (c) 2015-2022, EPFL/Blue Brain Project
+ * All rights reserved. Do not distribute without permission.
+ * Responsible Author: Nadir Roman Guerrero <nadir.romanguerrero@epfl.ch>
+ *
+ * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+#include <typeindex>
+#include <vector>
+
+namespace brayns
+{
+class Components
+{
+public:
+    template<typename T, typename... Args>
+    T &add(Args &&...args)
+    {
+        static_assert(std::is_same_v<std::decay_t<T>, T>, "Only decayed components allowed");
+
+        auto &entry = _components.emplace_back(typeid(T), std::make_unique<Component<T>>());
+        auto &component = static_cast<Component<T> &>(*entry.component);
+        auto &data = component.data;
+
+        if constexpr (std::is_aggregate_v<T>)
+        {
+            data = T{std::forward<Args>(args)...};
```

Comment:
This one works also for non-aggregates no ?

Reply:
Yes, but the purpose of the condition is to be able to pass directly struct fields in add()